javalang/parser.py

```diff
@@ -1731,7 +1740,23 @@ def parse_expressionl(self):
             return tree.TernaryExpression(condition=expression_2,
                                           if_true=true_expression,
                                           if_false=false_expression)
-
+        if self.would_accept('->'):
+            body = self.parse_lambda_method_body()
+            return tree.LambdaExpression(parameters=[expression_2],
+                                         body=body)
+        if self.try_accept('::'):
+            if self.would_accept('<'):
```

Comment:
This method reference code doesn't seem right.

- There may be multiple type arguments. I think it would be appropriate to use `parse_nonwildcard_type_arguments` here.
- Type argument shouldn't be represented using `tree.Cast`.
- I do not believe from reading the spec that a generic expression is accepted on the right of the '::',

```
MethodReference:
ExpressionName :: [TypeArguments] Identifier
ReferenceType :: [TypeArguments] Identifier
Primary :: [TypeArguments] Identifier
super :: [TypeArguments] Identifier
TypeName . super :: [TypeArguments] Identifier
ClassType :: [TypeArguments] new
ArrayType :: new
```

I don't have an immediate idea for how to correctly support all of theses cases, but I have a few recommendations,

- After the '::' accept optional type arguments (using the `parse_nonwildcard_type_arguments`) followed by a single Identifier or "new".
- Rename `ExpressionWithMemberReference` to `MethodReference` and add a `type_arguments` attr to it. You could also rename the `method_reference` attr to `method` to communicate that it is only the method name and not the complete reference.
- Construct tests cases for a few of the method reference types. I suspect there may be some issues with the `super` method references.